Zend/zend_interfaces.stub.php

```diff
@@ -49,3 +49,9 @@ interface Countable
     /** @return int */
     function count();
 }
+
+interface Stringable
+{
+    /** @return string */
```

Comment:
Can this internal type declare an explicit return type?

Reply:
I don't think we can, because `__toString` doesn't enforce the return type at declaration time (but it does at runtime, so this is still type-safe).

If the interface were made to enforce declaring the return type, it would make it quite hard for libraries/apps to adopt the polyfill, as they would be force to ... declare the return type, which would be a BC break.

Reply:
I strongly recommend enforcing a string type on the interface. This would provide additional value beyond a union, and helps shore up the result. Remember that we couldn't throw exceptions in `__toString` until recently but since we can now I would prefer that a magic method whose purpose is to convert to a string *always return a string*.

Reply:
The reason why I ask this is that it is a new interface, so it can indeed be introduced. Whether internal symbols then implement `Stringable` is more of a BC discussion.

I was more questioning whether core can now declare return types at all: there was a reason why this wasn't feasible before.

Reply:
I think PHP 8 might want to enforce the `string` return type on `__toString` before we can add it to the interface. See my concerns with BC above. Adding the return type to the interface but not to `__toString` might slow down its adoption a lot (because adopting it would mean breaking BC for libs)

Reply:
> I was more questioning whether core can now declare return types at all

Yes, it can. Most internal *functions* already have a return type declaration by now (e.g. https://github.com/php/php-src/blob/master/ext/standard/basic_functions.stub.php), while existing internal *methods* don't have one only because of BC.

Reply:
> Note that the method on the interface doesn't declare the string return type on __toString because that would require adding the return type to e.g. the Exception class, this breaking BC. And adding another BC break to PHP 8 is not the goal of this PR.

Added to the description of the PR.

Reply:
> 
> 
> > I was more questioning whether core can now declare return types at all
> 
> Yes, it can. Most internal _functions_ already have a return type declaration by now (e.g. https://github.com/php/php-src/blob/master/ext/standard/basic_functions.stub.php), while existing internal _methods_ only don't have one because of BC.

Internal *non final* methods, if I may specify.
I would also rather prefer to have this new interface declare a ``string`` return type and not have ``Throwable`` extend from it to preserve BC (as it's clearly know that the Throwable interface has the __toString method)

Reply:
`Throwable` is just one example. The BC concern happens with ever single class, internal or not, that implements `__toString()`.

Reply:
@nicolas-grekas but you would need to add the interface, meaning that people can work by making it a reasonable upgrade path. Sure you can't just drop in the interface without some small effort/planning. But this is better to have people do some effort NOW then have an interface which doesn't enforce the contract.

Reply:
Adding the interface can be done right away in PHP7 in a minor release of any exiting apps, with a very simple polyfill (even PHP5 but it that's not important).

Adding a return type would force a major release of all existing libs that want to adopt the interface, with all associated costs of a BC break.

As I mentioned before, this interface shouldn't be the place to add the return type declaration. What should be done instead is a separate RFC that proposes to enforce `:string` to the magic `__toString()` method. This would be the proper space do discuss if breaking BC here is worth it in PHP 8.

But adding the return type only to the interface without adding to the standalone magic method only defeats the usefulness of this interface, by adding a significant cost to its adoption.

Reply:
I see what you mean, but, it's not possible to add the return type to the magic ``__toString()`` as it would break 99% of code in existence as *all* code which doesn't add a return type to the ``__toString()`` will break, this is a massive BC break.

The fact to add the return type to new interface has no BC implication what so ever, sure implementing it may be a bit more annoying, but this is far better then breaking all PHP code.

This is only my opinion but if the interface doesn't declare a ``string`` return type I'll vote against it. As I don't think making it more complicate for people to implement a new interface hold much value compared to breaking vast amount of code.

Reply:
What about if the interface had the return type, classes without the return type that don't implement the interface still magically show up as implementing it? This would solve the BC issue, because people would only need to update their code to implement the interface syntactically

Reply:
Adding more magic to the engine is something we should avoid IMHO. We have enough semantics in the engine now to use plain boring union types, which is a good thing(c) !

Reply:
> classes without the return type that don't implement the interface still magically show up as implementing it

on same suggestion by @bwoebi in https://github.com/php/php-src/pull/5083#issuecomment-574223256, and because it solves the concern in this thread, and mines also, I'm going to follow this.

Now marking this thread as resolved.

Reply:
> classes without the return type that don't implement the interface still magically show up as implementing it

on same suggestion by @bwoebi in #5083 (comment), and because it solves the concern in this thread, and mines also, I'm going to follow this. (I figured out how to do it, update coming soon)

Now marking this thread as resolved.